File: scripts/lto/agent_job.py
"""AgentJob / AgentResult — agent 世界的数据合同（harness 地基）。

这是 spawn 原语(agent_exec)、调度器(scheduler)、事实简报器(lto next)
三方共享的单一真源。刻意区别于 exec.py 的 shell 世界合同
（command/cwd/rc/stdout/stderr）——agent harness 编排的是带独立 context 的
agent，不是 shell 命令。

设计原则（codex 异构审纠正）：
- 不用 run_command 的 shell 签名锁死 spawn，否则 host agent 的组合空间会被卡住。
- 每个字段都对应一个真实消费者：
  · scheduler 读 budget / retry_policy / runner（并发、退避、限流）
  · agent_exec 读 prompt_ref / runner / model / isolation / output_schema
  · lto next 读 parent_pattern / verifier_of / children（给 host agent 的编排事实）
"""

from __future__ import annotations

from dataclasses import dataclass, field, asdict
from enum import Enum
from typing import Any


# 编排 pattern（官方文章中 pattern 的实用子集，按真实需求增量铺开）
class Pattern(str, Enum):
    LINEAR = "linear"            # 单 agent 顺序
    FAN_OUT = "fan-out"          # 拆多步并发 + barrier 合成
    ADVERSARIAL = "adversarial"  # 每个 generator 配独立 verifier
    TOURNAMENT = "tournament"    # placeholder-only：枚举占位，无真实触发场景未实现（YAGNI）
    LOOP = "loop"                # placeholder-only：枚举占位，无真实触发场景未实现（YAGNI）


# 已知 runner 家族（与 audit.py 的家族映射保持一致，跨 runtime 异构判定用）
KNOWN_RUNNERS = ("codex", "pi", "agy", "gemini", "claude")
CODEX_SANDBOXES = ("read-only", "workspace-write", "danger-full-access")

# 跨 runner read-only 兑现机制（见 references/runner-readonly-contract.md）。
# 四家机制不可通约——codex 是 sandbox 三级，agy 是 sandbox 开关(=workspace-write，
# 无 read-only 档)，pi/claude 是工具白名单（pi 替换语义真裁、claude 靠 plan mode 软约束）。
# 这些 allowlist 是 §2.1 的只读安全工具精确集，大小写敏感。
READONLY_TOOL_ALLOWLIST = {
    "claude": frozenset({"Read", "Grep", "Glob", "WebFetch"}),
    "pi": frozenset({"read", "grep", "find", "ls"}),
}


class JobStatus(str, Enum):
    PENDING = "pending"
    RUNNING = "running"
    OK = "ok"
    FAILED = "failed"
    TIMEOUT = "timeout"
    RATE_LIMITED = "rate_limited"
    SKIPPED = "skipped"


@dataclass
class PermissionPolicy:
    """Per-job permission intent, used by scheduler/runner env guards.

    sandbox 是归一化的跨 runner 等级（read-only/workspace-write/danger-full-access）。
    tools 是 tool-allowlist 机制(pi/claude)下 scheduler 构造侧要注入的工具集——
    pi M-v2：用结构化字段而非 ad-hoc env 字典，避免绕类型检查。非 tool 机制留空。
    """
    sandbox: str = "read-only"
    reason: str = ""
    user_approved: bool = False
    tools: tuple[str, ...] = ()  # tool-allowlist runner 注入的工具集（pi/claude）

    def validate_for_runner(self, runner: str, env: dict[str, str]) -> None:
        if not all(isinstance(k, str) and isinstance(v, str) for k, v in env.items()):
            raise ValueError("env keys and values must be strings")
        if self.sandbox not in CODEX_SANDBOXES:
            raise ValueError(f"invalid sandbox: {self.sandbox!r}")

        # 通用越权前置约束（任何 runner、任何机制）：高权限需 reason / 审批。
        if self.sandbox == "workspace-write" and not self.reason.strip():
            raise ValueError("workspace-write requires permission_policy.reason")
        if self.sandbox == "danger-full-access":
            if not self.reason.strip():
                raise ValueError("danger-full-access requires permission_policy.reason")
            if not self.user_approved:
                raise ValueError("danger-full-access requires user_approved=True")

        if runner == "codex":
            # codex 用 env 通道（CODEX_SANDBOX），必须与 policy 一致（不双路径）。
            sandbox = env.get("CODEX_SANDBOX", self.sandbox)
            if sandbox != self.sandbox:
                raise ValueError(
                    "CODEX_SANDBOX conflicts with permission_policy.sandbox "
                    f"({sandbox!r} != {self.sandbox!r})"
                )
            return

        if runner in ("agy", "gemini"):
            # §7 实测：agy --sandbox = workspace-write，无 read-only 档。
            # read-only profile 下 agy 无可兑现档位 → validate 阶段即拒绝（item 7）。
            # gemini 同样无任何 read-only 兑现机制（gemini.sh 裸跑），却曾被本分支
            # 漏掉直接放行——2026-06-10 三方 spec 审（agy）审出，同等 fail-closed。
            if self.sandbox == "read-only":
                raise ValueError(
                    f"{runner} cannot enforce read-only (no read-only sandbox "
                    f"mechanism); defer {runner} for read-only jobs"
                )
            return

        if runner in READONLY_TOOL_ALLOWLIST:
            # claude / pi：tool-allowlist 机制。read-only 时 tools 必须是只读安全子集。
            # 空 tools 合法——scheduler 构造侧会注入该 runner 的默认只读集
            # (effective_readonly_tools)，零回归不强迫调用方到处显式传。fail-closed
            # 的兜底在运行后的 _sandbox_exceeds（按 actual_tools 判），不在此前置拒绝。
            # 只在调用方显式给了 allowlist 外的工具时才拒（明确越读意图）。
            if self.sandbox == "read-only" and self.tools:
                allow = READONLY_TOOL_ALLOWLIST[runner]
                extra = set(self.tools) - allow
                if extra:
                    raise ValueError(
                        f"{runner} read-only tools exceed allowlist: {sorted(extra)} "
                        f"not in {sorted(allow)}"
                    )
            return

    def effective_readonly_tools(self, runner: str) -> tuple[str, ...]:
        """scheduler 构造侧用：read-only + tool-allowlist runner 的实际注入工具集。

        显式 tools 优先；否则回落到该 runner 的标准只读 allowlist（§2.1）。
        非 read-only 或非 tool-allowlist runner 返回空。
        """
        if self.sandbox != "read-only" or runner not in READONLY_TOOL_ALLOWLIST:
            return ()
        if self.tools:
            return self.tools
        return tuple(sorted(READONLY_TOOL_ALLOWLIST[runner]))

    def to_dict(self) -> dict[str, Any]:
        d = asdict(self)
        d["tools"] = list(self.tools)
        return d


@dataclass
class Budget:
    """单 job 资源预算。scheduler 据此限流与计量。"""
    timeout_sec: int = 300
    max_tokens: int | None = None      # 可选 token 上限（成本计量）

    def to_dict(self) -> dict[str, Any]:
        return asdict(self)


@dataclass
class RetryPolicy:
    """重试 + 退避。scheduler 在 rate_limited / timeout 时据此重试。"""
    max_retries: int = 1
    backoff_sec: float = 5.0           # 首次退避，指数增长
    retry_on: tuple[str, ...] = ("rate_limited", "timeout")

    def to_dict(self) -> dict[str, Any]:
        return {**asdict(self), "retry_on": list(self.retry_on)}


@dataclass
class AgentJob:
    """一个待调度的 agent 任务（host-agent-facing 合同）。"""
    job_id: str
    prompt_ref: str                                # 简报文件路径或内联文本
    runner: str = "codex"                          # KNOWN_RUNNERS 之一
    prompt_is_inline: bool = False                 # True=prompt_ref 是文本不是路径
    model: str | None = None                       # 可选，model routing 用
    env: dict[str, str] = field(default_factory=dict)  # per-job runner env (CODEX_SANDBOX, etc.)
    permission_policy: PermissionPolicy = field(default_factory=PermissionPolicy)
    isolation: str = "none"                        # none | worktree
    output_schema: dict[str, Any] | None = None    # 强制审者结构化输出（findings/severity）
    parent_pattern: str = Pattern.LINEAR.value     # 本 job 属于哪个编排 pattern
    budget: Budget = field(default_factory=Budget)
    retry_policy: RetryPolicy = field(default_factory=RetryPolicy)
    verifier_of: str | None = None                 # 若本 job 对抗验证某 job，指向其 job_id
    children: list[str] = field(default_factory=list)  # fan-out 子 job_id
    meta: dict[str, Any] = field(default_factory=dict)  # 自由扩展（task_id 等）

    def validate(self) -> None:
        if self.runner not in KNOWN_RUNNERS:
            raise ValueError(f"unknown runner: {self.runner!r} (known: {KNOWN_RUNNERS})")
        if self.isolation not in ("none", "worktree"):
            raise ValueError(f"invalid isolation: {self.isolation!r}")
        valid_patterns = {p.value for p in Pattern}
        if self.parent_pattern not in valid_patterns:
            raise ValueError(f"invalid parent_pattern: {self.parent_pattern!r}")
        if not self.prompt_ref:
            raise ValueError("prompt_ref is required")
        self.permission_policy.validate_for_runner(self.runner, self.env)

    def to_dict(self) -> dict[str, Any]:
        d = asdict(self)
        d["budget"] = self.budget.to_dict()
        d["retry_policy"] = self.retry_policy.to_dict()
        d["permission_policy"] = self.permission_policy.to_dict()
        return d

    @classmethod
    def from_dict(cls, d: dict[str, Any]) -> "AgentJob":
        d = dict(d)
        if "budget" in d and isinstance(d["budget"], dict):
            d["budget"] = Budget(**d["budget"])
        if "retry_policy" in d and isinstance(d["retry_policy"], dict):
            rp = dict(d["retry_policy"])
            if "retry_on" in rp and isinstance(rp["retry_on"], list):
                rp["retry_on"] = tuple(rp["retry_on"])
            d["retry_policy"] = RetryPolicy(**rp)
        if "permission_policy" in d and isinstance(d["permission_policy"], dict):
            pp = dict(d["permission_policy"])
            if "tools" in pp and isinstance(pp["tools"], list):
                pp["tools"] = tuple(pp["tools"])
            d["permission_policy"] = PermissionPolicy(**pp)
        known = {f for f in cls.__dataclass_fields__}
        return cls(**{k: v for k, v in d.items() if k in known})


@dataclass
class AgentResult:
    """一个 agent 任务的执行结果。"""
    job_id: str
    runner: str
    model: str | None = None                       # 具体模型型号（runner 是家族，model 是型号；跨 run 挖掘按 model 子分组用）
    status: str = JobStatus.PENDING.value
    exit_code: int | None = None                   # 区分 124 timeout / 0 空返回 / 非零失败
    findings: list[dict[str, Any]] = field(default_factory=list)  # 结构化（非 regex 扫文本）
    reply_text: str = ""                           # 原始回复（findings 解析失败时兜底）
    cost: dict[str, Any] = field(default_factory=dict)  # tokens / elapsed_sec
    permissions: dict[str, Any] = field(default_factory=dict)  # runner sandbox/env decision snapshot
    artifacts: list[str] = field(default_factory=list)  # reply 文件路径等
    attempts: int = 1                              # 实际尝试次数（含重试）
    error: str = ""                                # 失败时的诊断信息

    @property
    def ok(self) -> bool:
        return self.status == JobStatus.OK.value

    def severity_counts(self) -> dict[str, int]:
        """从结构化 findings 数 severity（P1 去 regex 的基础）。"""
        counts = {"critical": 0, "high": 0, "medium": 0, "low": 0}
        for f in self.findings:
            sev = str(f.get("severity", "")).lower()
            if sev in counts:
                counts[sev] += 1
        return counts

    def to_dict(self) -> dict[str, Any]:
        return asdict(self)

    @classmethod
    def from_dict(cls, d: dict[str, Any]) -> "AgentResult":
        known = {f for f in cls.__dataclass_fields__}
        return cls(**{k: v for k, v in d.items() if k in known})

File: src/agent_job.rs
use serde::{Deserialize, Serialize};
use std::collections::BTreeMap;
use std::fmt;
use std::str::FromStr;
use thiserror::Error;

pub const KNOWN_RUNNERS: &[&str] = &["codex", "pi", "agy", "gemini", "claude"];
pub const CODEX_SANDBOXES: &[&str] = &["read-only", "workspace-write", "danger-full-access"];

#[derive(Debug, Error, PartialEq, Eq)]
pub enum AgentJobError {
    #[error("unknown runner: {0}")]
    UnknownRunner(String),
    #[error("invalid sandbox: {0}")]
    InvalidSandbox(String),
    #[error("workspace-write requires permission_policy.reason")]
    MissingWorkspaceWriteReason,
    #[error("danger-full-access requires permission_policy.reason")]
    MissingDangerReason,
    #[error("danger-full-access requires user_approved=true")]
    MissingDangerApproval,
    #[error("CODEX_SANDBOX conflicts with permission_policy.sandbox ({actual} != {expected})")]
    CodexSandboxConflict { actual: String, expected: String },
    #[error("{runner} cannot enforce read-only; defer it for read-only jobs")]
    CannotEnforceReadOnly { runner: String },
    #[error("{runner} read-only tools exceed allowlist: {extra:?}")]
    ReadOnlyToolsExceeded { runner: String, extra: Vec<String> },
    #[error("prompt_ref is required")]
    MissingPromptRef,
    #[error("invalid isolation: {0}")]
    InvalidIsolation(String),
    #[error("invalid parent_pattern: {0}")]
    InvalidPattern(String),
    #[error("env keys and values must be strings")]
    InvalidEnv,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum Sandbox {
    ReadOnly,
    WorkspaceWrite,
    DangerFullAccess,
}

impl Sandbox {
    pub fn as_str(self) -> &'static str {
        match self {
            Self::ReadOnly => "read-only",
            Self::WorkspaceWrite => "workspace-write",
            Self::DangerFullAccess => "danger-full-access",
        }
    }
}

impl fmt::Display for Sandbox {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.write_str(self.as_str())
    }
}

impl FromStr for Sandbox {
    type Err = AgentJobError;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s {
            "read-only" => Ok(Self::ReadOnly),
            "workspace-write" => Ok(Self::WorkspaceWrite),
            "danger-full-access" => Ok(Self::DangerFullAccess),
            other => Err(AgentJobError::InvalidSandbox(other.to_string())),
        }
    }
}

#[derive(Debug, Clone, Copy, Default, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum TaskSize {
    Small,
    Medium,
    Large,
    #[default]
    Unknown,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum JobStatus {
    Pending,
    Running,
    Ok,
    Failed,
    Timeout,
    RateLimited,
    Skipped,
}

impl JobStatus {
    pub fn as_str(self) -> &'static str {
        match self {
            Self::Pending => "pending",
            Self::Running => "running",
            Self::Ok => "ok",
            Self::Failed => "failed",
            Self::Timeout => "timeout",
            Self::RateLimited => "rate_limited",
            Self::Skipped => "skipped",
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ExitState {
    Ok(i32),
    Failed(i32),
    SignalKilled,
    Timeout,
}

impl ExitState {
    pub fn from_status_code(code: Option<i32>, timed_out: bool) -> Self {
        if timed_out {
            return Self::Timeout;
        }
        match code {
            Some(0) => Self::Ok(0),
            Some(124) => Self::Timeout,
            Some(code) => Self::Failed(code),
            None => Self::SignalKilled,
        }
    }
}

#[derive(Debug, Clone, Default, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum Pattern {
    #[default]
    Linear,
    FanOut,
    Adversarial,
    Tournament,
    Loop,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, Default)]
pub struct Usage {
    pub tokens_in: Option<u64>,
    pub tokens_out: Option<u64>,
    pub tokens: Option<u64>,
    #[serde(flatten)]
    pub extra: BTreeMap<String, serde_json::Value>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct PermissionPolicy {
    pub sandbox: Sandbox,
    #[serde(default)]
    pub reason: String,
    #[serde(default)]
    pub user_approved: bool,
    #[serde(default)]
    pub tools: Vec<String>,
}

impl Default for PermissionPolicy {
    fn default() -> Self {
        Self {
            sandbox: Sandbox::ReadOnly,
            reason: String::new(),
            user_approved: false,
            tools: Vec::new(),
        }
    }
}

impl PermissionPolicy {
    pub fn validate_for_runner(
        &self,
        runner: &str,
        env: &BTreeMap<String, String>,
    ) -> Result<(), AgentJobError> {
        if !KNOWN_RUNNERS.contains(&runner) {
            return Err(AgentJobError::UnknownRunner(runner.to_string()));
        }

        match self.sandbox {
            Sandbox::ReadOnly => {}
            Sandbox::WorkspaceWrite if self.reason.trim().is_empty() => {
                return Err(AgentJobError::MissingWorkspaceWriteReason);
            }
            Sandbox::WorkspaceWrite => {}
            Sandbox::DangerFullAccess if self.reason.trim().is_empty() => {
                return Err(AgentJobError::MissingDangerReason);
            }
            Sandbox::DangerFullAccess if !self.user_approved => {
                return Err(AgentJobError::MissingDangerApproval);
            }
            Sandbox::DangerFullAccess => {}
        }

        match runner {
            "codex" => {
                let actual = env
                    .get("CODEX_SANDBOX")
                    .map(String::as_str)
                    .unwrap_or_else(|| self.sandbox.as_str());
                if actual != self.sandbox.as_str() {
                    return Err(AgentJobError::CodexSandboxConflict {
                        actual: actual.to_string(),
                        expected: self.sandbox.to_string(),
                    });
                }
            }
            "agy" | "gemini" if self.sandbox == Sandbox::ReadOnly => {
                return Err(AgentJobError::CannotEnforceReadOnly {
                    runner: runner.to_string(),
                });
            }
            "claude" | "pi" if self.sandbox == Sandbox::ReadOnly && !self.tools.is_empty() => {
                let allow = readonly_tool_allowlist(runner);
                let mut extra = self
                    .tools
                    .iter()
                    .filter(|tool| !allow.contains(&tool.as_str()))
                    .cloned()
                    .collect::<Vec<_>>();
                extra.sort();
                if !extra.is_empty() {
                    return Err(AgentJobError::ReadOnlyToolsExceeded {
                        runner: runner.to_string(),
                        extra,
                    });
                }
            }
            _ => {}
        }
        Ok(())
    }

    pub fn effective_readonly_tools(&self, runner: &str) -> Vec<String> {
        if self.sandbox != Sandbox::ReadOnly || !matches!(runner, "claude" | "pi") {
            return Vec::new();
        }
        if !self.tools.is_empty() {
            return self.tools.clone();
        }
        readonly_tool_allowlist(runner)
            .iter()
            .map(|tool| (*tool).to_string())
            .collect()
    }
}

pub fn readonly_tool_allowlist(runner: &str) -> &'static [&'static str] {
    match runner {
        "claude" => &["Glob", "Grep", "Read", "WebFetch"],
        "pi" => &["find", "grep", "ls", "read"],
        _ => &[],
    }
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct Budget {
    #[serde(default = "default_timeout")]
    pub timeout_sec: u64,
    #[serde(default)]
    pub max_tokens: Option<u64>,
}

fn default_timeout() -> u64 {
    300
}

impl Default for Budget {
    fn default() -> Self {
        Self {
            timeout_sec: default_timeout(),
            max_tokens: None,
        }
    }
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct RetryPolicy {
    #[serde(default = "default_max_retries")]
    pub max_retries: u32,
    #[serde(default = "default_backoff")]
    pub backoff_sec: f64,
    #[serde(default = "default_retry_on")]
    pub retry_on: Vec<JobStatus>,
}

fn default_max_retries() -> u32 {
    1
}

fn default_backoff() -> f64 {
    5.0
}

fn default_retry_on() -> Vec<JobStatus> {
    vec![JobStatus::RateLimited, JobStatus::Timeout]
}

impl Default for RetryPolicy {
    fn default() -> Self {
        Self {
            max_retries: default_max_retries(),
            backoff_sec: default_backoff(),
            retry_on: default_retry_on(),
        }
    }
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct AgentJob {
    pub job_id: String,
    pub prompt_ref: String,
    #[serde(default = "default_runner")]
    pub runner: String,
    #[serde(default)]
    pub prompt_is_inline: bool,
    #[serde(default)]
    pub model: Option<String>,
    #[serde(default)]
    pub env: BTreeMap<String, String>,
    #[serde(default)]
    pub permission_policy: PermissionPolicy,
    #[serde(default = "default_isolation")]
    pub isolation: String,
    #[serde(default)]
    pub output_schema: Option<serde_json::Value>,
    #[serde(default)]
    pub parent_pattern: Pattern,
    #[serde(default)]
    pub budget: Budget,
    #[serde(default)]
    pub retry_policy: RetryPolicy,
    #[serde(default)]
    pub verifier_of: Option<String>,
    #[serde(default)]
    pub children: Vec<String>,
    #[serde(default)]
    pub task_type: Option<String>,
    #[serde(default)]
    pub size: TaskSize,
    #[serde(default)]
    pub meta: BTreeMap<String, serde_json::Value>,
}

fn default_runner() -> String {
    "codex".to_string()
}

fn default_isolation() -> String {
    "none".to_string()
}

impl AgentJob {
    pub fn validate(&self) -> Result<(), AgentJobError> {
        if !KNOWN_RUNNERS.contains(&self.runner.as_str()) {
            return Err(AgentJobError::UnknownRunner(self.runner.clone()));
        }
        if !matches!(self.isolation.as_str(), "none" | "worktree") {
            return Err(AgentJobError::InvalidIsolation(self.isolation.clone()));
        }
        if self.prompt_ref.is_empty() {
            return Err(AgentJobError::MissingPromptRef);
        }
        self.permission_policy
            .validate_for_runner(&self.runner, &self.env)
    }
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct AgentResult {
    pub job_id: String,
    pub runner: String,
    #[serde(default)]
    pub model: Option<String>,
    #[serde(default = "default_status")]
    pub status: JobStatus,
    #[serde(default)]
    pub exit_code: Option<i32>,
    #[serde(default)]
    pub findings: Vec<serde_json::Value>,
    #[serde(default)]
    pub reply_text: String,
    #[serde(default)]
    pub cost: BTreeMap<String, serde_json::Value>,
    #[serde(default)]
    pub permissions: BTreeMap<String, serde_json::Value>,
    #[serde(default)]
    pub artifacts: Vec<String>,
    #[serde(default = "default_attempts")]
    pub attempts: u32,
    #[serde(default)]
    pub error: String,
    #[serde(default)]
    pub task_type: Option<String>,
    #[serde(default)]
    pub size: TaskSize,
}

fn default_status() -> JobStatus {
    JobStatus::Pending
}

fn default_attempts() -> u32 {
    1
}

impl AgentResult {
    pub fn ok(&self) -> bool {
        self.status == JobStatus::Ok
    }

    pub fn severity_counts(&self) -> BTreeMap<String, usize> {
        let mut counts = BTreeMap::from([
            ("critical".to_string(), 0),
            ("high".to_string(), 0),
            ("medium".to_string(), 0),
            ("low".to_string(), 0),
        ]);
        for finding in &self.findings {
            if let Some(sev) = finding.get("severity").and_then(|v| v.as_str()) {
                let key = sev.to_ascii_lowercase();
                if let Some(count) = counts.get_mut(&key) {
                    *count += 1;
                }
            }
        }
        counts
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn codex_sandbox_env_must_match_policy() {
        let policy = PermissionPolicy::default();
        let env = BTreeMap::from([("CODEX_SANDBOX".to_string(), "workspace-write".to_string())]);
        assert!(matches!(
            policy.validate_for_runner("codex", &env),
            Err(AgentJobError::CodexSandboxConflict { .. })
        ));
    }

    #[test]
    fn agy_readonly_fails_closed() {
        let policy = PermissionPolicy::default();
        assert!(matches!(
            policy.validate_for_runner("agy", &BTreeMap::new()),
            Err(AgentJobError::CannotEnforceReadOnly { .. })
        ));
    }

    #[test]
    fn danger_full_requires_reason_and_approval() {
        let policy = PermissionPolicy {
            sandbox: Sandbox::DangerFullAccess,
            ..PermissionPolicy::default()
        };
        assert_eq!(
            policy.validate_for_runner("codex", &BTreeMap::new()),
            Err(AgentJobError::MissingDangerReason)
        );
        let policy = PermissionPolicy {
            sandbox: Sandbox::DangerFullAccess,
            reason: "release publish".to_string(),
            user_approved: false,
            tools: Vec::new(),
        };
        assert_eq!(
            policy.validate_for_runner("codex", &BTreeMap::new()),
            Err(AgentJobError::MissingDangerApproval)
        );
    }

    #[test]
    fn signal_killed_is_not_empty_reply_or_timeout() {
        assert_eq!(
            ExitState::from_status_code(None, false),
            ExitState::SignalKilled
        );
        assert_eq!(
            ExitState::from_status_code(Some(124), false),
            ExitState::Timeout
        );
        assert_eq!(
            ExitState::from_status_code(Some(1), false),
            ExitState::Failed(1)
        );
    }
}
